api/app/src/routes/medical/link.ts

```diff
@@ -1,56 +1,100 @@
-import { asyncHandler, getCxIdOrFail, getPatientIdFromQueryOrFail } from "../util";
-const router = Router();
 import status from "http-status";
+import {
+  asyncHandler,
+  getCxIdOrFail,
+  getPatientIdFromQueryOrFail,
+  getEntityIdFromQueryOrFail,
+  getLinkIdFromQueryOrFail,
+} from "../util";
+const router = Router();
 import { getPatient } from "../../command/medical/patient/get-patient";
+import { updatePatient } from "../../command/medical/patient/update-patient";
+import {
+  getPersonsAtCommonwell,
+  getLinkFromCommonwell,
+  linkPatientToCommonwellPerson,
+  resetCommonwellLink,
+} from "../../external/commonwell/link";
+import { Config } from "../../shared/config";
+import { PatientLinks } from "./schemas/link";
+import { LinkSource } from "./schemas/link";
 
 /** ---------------------------------------------------------------------------
  * POST /link
  *
  * Creates link to the specified entity.
+ * @param   req.query.patientId     Patient ID to link to a person.
+ * @param   req.query.entityId      Person ID to link to the patient.
+ * @param   req.query.linkSource    HIE from where the link is made too.
  *
- * @return  {Link}  The created link.
+ * @return  status 200. Let user know the link has been made.
  */
 router.post(
   "/",
   asyncHandler(async (req: Request, res: Response) => {
     const cxId = getCxIdOrFail(req);
     const patientId = getPatientIdFromQueryOrFail(req);
+    const entityId = getEntityIdFromQueryOrFail(req);
+
     const patient = await getPatient({ id: patientId, cxId });
-    // get linkSource from query params -> this should be "CommonWell"
-    console.log(patient);
+    const linkSource = req.query.linkSource;
 
-    // CommonWell.patientLink();
-    // store the following data in our DB in the patient:
-    //    - link ID
-    //    - person ID
-    // can put this into a JSONB column `link_data` with structure:
-    //    {["CommonWell"]: {cw_link_id, cw_person_id}}
+    // TODO: HANDLE OTHER HIE's
+    if (linkSource === LinkSource.commonWell) {
+      const linkId = await linkPatientToCommonwellPerson(entityId, patient.patientNumber);
 
-    return res.status(status.OK).json({});
+      await updatePatient({
+        id: patientId,
+        cxId,
+        linkData: {
+          ...patient.linkData,
+          [linkSource]: {
+            cw_link_id: linkId,
+            cw_person_id: entityId,
+          },
+        },
+      });
+      return res.status(status.OK).json(linkId);
+    }
+
+    throw new Error("Link source not found");
   })
 );
 
 /** ---------------------------------------------------------------------------
  * DELETE /link
  *
  * Removes the specified HIE link from the specified patient.
- *
+ * @param   req.query.patientId     Patient ID to remove link from.
+ * @param   req.query.linkId        Link ID to remove.
+ * @param   req.query.linkSource    HIE to remove the link from.
  * @return  200
  */
 router.delete(
   "/",
```

Comment:
as per @leite08's comments regarding REST design standards, the `id` of the resource should be in the URL & not the query params, ie `/:linkId` for all requests aside from the `POST` that makes it -> I didn't know that this was a standard myself until this was brought up

Reply:
That makes sense but in this case the linkId for commonwell is an href so that wouldn't work in this case.

`https://integration.rest.api.commonwellalliance.org/v1/person/6f5338aa-d8e6-4199-bd5c-21879507b87c/patientLink/1000000002%5E%5E%5Eurn%3aoid%3a2.16.840.1.113883.3.9621/`

 A link is pretty much the person and patientid

Reply:
Yeah, this is one of the drawbacks of not having our representation of the whole domain. We're exposing an underlying dependency and are impacted by how it works.

Agree there's nothing much we can do here... could represent the link internally as base64 of the CW link, this should help... we would convert it to/from base64 an `external/commonwell/*.ts` file, abstracting it away from the API clients.